src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -160,35 +175,443 @@ private DeclaredTypeAssignment GetParameterType(ParameterDeclarationSyntax synta
+
+            if (type is ResourceType resourceType && IsExtensibilityType(resourceType))
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).UnsupportedResourceTypeParameterOrOutputType(resourceType.Name));
+            }
+
+            return type;
+        }
+
+        private bool IsExtensibilityType(ResourceType resourceType)
+        {
+            return resourceType.DeclaringNamespace.ProviderName != AzNamespaceType.BuiltInName;
+        }
+
+        private TypeSymbol? GetOutputValueType(SyntaxBase syntax) => binder.GetParent(syntax) switch
+        {
+            OutputDeclarationSyntax outputDeclaration => typeManager.GetTypeInfo(outputDeclaration.Value),
+            ParenthesizedExpressionSyntax parenthesized => GetOutputValueType(parenthesized),
+            _ => null,
+        };
+
+        private TypeSymbol ParseTypeExpression(TypeAccessSyntax syntax)
+            => binder.GetSymbolInfo(syntax) switch
+            {
+                UserDefinedTypeSymbol declaredType => TypeRefToType(syntax, declaredType),
+                DeclaredSymbol declaredSymbol => ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).ValueSymbolUsedAsType(declaredSymbol.Name)),
+                _ => ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).SymbolicNameIsNotAType(syntax.Name.IdentifierName, GetValidTypeNames())),
+            };
+
+        private IEnumerable<string> GetValidTypeNames() => LanguageConstants.DeclarationTypes.Keys
+                .Concat(binder.FileSymbol.TypeDeclarations.Select(td => td.Name))
+                .Distinct();
+
+        private TypeSymbol TypeRefToType(TypeAccessSyntax signifier, UserDefinedTypeSymbol signified)
+        {
+            var signifiedType = userDefinedTypeReferences.GetOrAdd(signified, GetUserDefinedTypeType);
+            if (signifiedType is ErrorType error)
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(signifier).ReferencedSymbolHasErrors(signified.Name));
+            }
+
+            return signifiedType;
+        }
+
+        private TypeSymbol ParseTypeExpression(ArrayTypeSyntax syntax)
+        {
+            if (!features.UserDefinedTypesEnabled)
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).TypedArrayDeclarationsUnsupported());
+            }
+
+            return new TypedArrayType(GetTypeFromTypeSyntax(syntax.Item), TypeSymbolValidationFlags.Default);
+        }
+
+        private TypeSymbol ParseTypeExpression(ObjectTypeSyntax syntax)
+        {
+            if (!features.UserDefinedTypesEnabled)
+            {
+                return ErrorType.Create(DiagnosticBuilder.ForPosition(syntax).TypedObjectDeclarationsUnsupported());
+            }
+
+            List<TypeProperty> properties = new();
+            List<ErrorDiagnostic> diagnostics = new();
+            StringBuilder objectName = new("{ ");
```

Comment:
Could we separate out logic for generating anonymous type names into a static helper method? (`TypeHelper` feels like a good candidate as it's already doing something similar for union types)